File: core/src/test/java/de/qabel/core/drop/DropMessageGsonTest.java
package de.qabel.core.drop;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import com.google.gson.JsonSyntaxException;
import de.qabel.core.config.Contact;
import de.qabel.core.config.Identities;
import de.qabel.core.config.Identity;
import de.qabel.core.config.factory.DropUrlGenerator;
import de.qabel.core.crypto.QblECKeyPair;
import org.junit.Test;

import java.util.Arrays;
import java.util.Collections;

import static org.junit.Assert.*;

public class DropMessageGsonTest {

    public static final String TEST_MESSAGE = "baz";
    public static final String TEST_MESSAGE_TYPE = "test_message";

    @Test(expected = JsonSyntaxException.class)
    public void invalidJsonDeserializeTest() {
        // 'time' got a wrong value
        String json = "{\"version\":1,\"time\":\"asdf\",\"sender\":\"foo\",\"acknowledgeID\":\"1234\",\"model\":\"DropMessageGsonTest$TestMessage\",\"data\":\"{\\\"content\\\":\\\"bar\\\"}\"}";

        Gson gson = DropMessageGson.INSTANCE.create();

        gson.fromJson(json, DropMessage.class);
    }

    @Test(expected = JsonSyntaxException.class)
    public void missingAcknowledgeId() {
        String json = "{\"version\":1,\"time\":\"asdf\",\"sender\":\"foo\",\"model\":\"DropMessageGsonTest$TestMessage\",\"data\":\"{\\\"content\\\":\\\"bar\\\"}\"}";

        Gson gson = DropMessageGson.INSTANCE.create();

        gson.fromJson(json, DropMessage.class);
    }

    @Test(expected = JsonSyntaxException.class)
    public void invalidJsonDeserializeTest2() {
        // 'time' got a missing value, i.e. wrong syntax
        String json = "{\"version\":1,\"time\":,\"sender\":\"foo\",\"acknowledgeID\":\"1234\",\"model\":\"DropMessageGsonTest$TestMessage\",\"data\":\"{\\\"content\\\":\\\"bar\\\"}\"}";

        Gson gson = DropMessageGson.INSTANCE.create();

        gson.fromJson(json, DropMessage.class);
    }

    @Test
    public void serializeTest() throws Exception {
        DropURL dropURL = new DropUrlGenerator("http://drop.qabel.de").generateUrl();
        Identity sender = new Identity("Bernd", Collections.singletonList(dropURL), new QblECKeyPair());
        Identities identities = new Identities();
        identities.put(sender);
        DropMessage dropMessage = new DropMessage(sender, TEST_MESSAGE, TEST_MESSAGE_TYPE);

        Gson gson = DropMessageGson.INSTANCE.create();
        String json = gson.toJson(dropMessage);
        assertNotNull(json);

        DropMessage receivedDropMessage = gson.fromJson(json, DropMessage.class);

        assertTrue(receivedDropMessage.registerSender(sender));
        assertEquals(TEST_MESSAGE, receivedDropMessage.getDropPayload());
        assertEquals(sender, receivedDropMessage.getSender());
        assertEquals(dropMessage.getAcknowledgeID(), receivedDropMessage.getAcknowledgeID());
        assertEquals(dropMessage.getDropMessageMetadata(), receivedDropMessage.getDropMessageMetadata());
    }

    @Test
    public void testSerializeWithoutHello() throws Exception {
        DropURL dropURL = new DropUrlGenerator("http://drop.qabel.de").generateUrl();
        Contact sender = new Contact("Bernd", Collections.singletonList(dropURL), new QblECKeyPair().getPub());

        DropMessage dropMessage = new DropMessage(sender, TEST_MESSAGE, TEST_MESSAGE_TYPE);
        assertNull(dropMessage.getDropMessageMetadata());

        Gson gson = DropMessageGson.INSTANCE.create();
        String json = gson.toJson(dropMessage);
        assertNotNull(json);

        DropMessage receivedDropMessage = gson.fromJson(json, DropMessage.class);
        assertNull(receivedDropMessage.getDropMessageMetadata());
        assertEquals(sender.getKeyIdentifier(), receivedDropMessage.getSenderKeyId());
        assertEquals(dropMessage.getDropPayload(), receivedDropMessage.getDropPayload());
        assertEquals(dropMessage.getDropPayloadType(), receivedDropMessage.getDropPayloadType());
    }
}

File: core/src/test/java/de/qabel/core/drop/DropMessageGsonTest.kt
package de.qabel.core.drop

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonSyntaxException
import de.qabel.core.config.Contact
import de.qabel.core.config.Identities
import de.qabel.core.config.Identity
import de.qabel.core.config.factory.DropUrlGenerator
import de.qabel.core.crypto.QblECKeyPair
import org.junit.Test

import java.util.Arrays
import java.util.Collections

import org.junit.Assert.*

class DropMessageGsonTest {

    @Test(expected = JsonSyntaxException::class)
    fun invalidJsonDeserializeTest() {
        // 'time' got a wrong value
        val json = "{\"version\":1,\"time\":\"asdf\",\"sender\":\"foo\",\"acknowledgeID\":\"1234\",\"model\":\"DropMessageGsonTest\$TestMessage\",\"data\":\"{\\\"content\\\":\\\"bar\\\"}\"}"

        val gson = DropMessageGson.create()

        gson.fromJson(json, DropMessage::class.java)
    }

    @Test(expected = JsonSyntaxException::class)
    fun missingAcknowledgeId() {
        val json = "{\"version\":1,\"time\":\"asdf\",\"sender\":\"foo\",\"model\":\"DropMessageGsonTest\$TestMessage\",\"data\":\"{\\\"content\\\":\\\"bar\\\"}\"}"

        val gson = DropMessageGson.create()

        gson.fromJson(json, DropMessage::class.java)
    }

    @Test(expected = JsonSyntaxException::class)
    fun invalidJsonDeserializeTest2() {
        // 'time' got a missing value, i.e. wrong syntax
        val json = "{\"version\":1,\"time\":,\"sender\":\"foo\",\"acknowledgeID\":\"1234\",\"model\":\"DropMessageGsonTest\$TestMessage\",\"data\":\"{\\\"content\\\":\\\"bar\\\"}\"}"

        val gson = DropMessageGson.create()

        gson.fromJson(json, DropMessage::class.java)
    }

    @Test
    @Throws(Exception::class)
    fun serializeTest() {
        val dropURL = DropUrlGenerator("http://drop.qabel.de").generateUrl()
        val sender = Identity("Bernd", listOf(dropURL), QblECKeyPair())
        val identities = Identities()
        identities.put(sender)
        val dropMessage = DropMessage(sender, TEST_MESSAGE, TEST_MESSAGE_TYPE)

        val gson = DropMessageGson.create()
        val json = gson.toJson(dropMessage)
        assertNotNull(json)

        val receivedDropMessage = gson.fromJson(json, DropMessage::class.java)

        assertTrue(receivedDropMessage.registerSender(sender))
        assertEquals(TEST_MESSAGE, receivedDropMessage.dropPayload)
        assertEquals(sender, receivedDropMessage.sender)
        assertEquals(dropMessage.acknowledgeID, receivedDropMessage.acknowledgeID)
        assertEquals(dropMessage.dropMessageMetadata, receivedDropMessage.dropMessageMetadata)
    }

    @Test
    @Throws(Exception::class)
    fun testSerializeWithoutHello() {
        val dropURL = DropUrlGenerator("http://drop.qabel.de").generateUrl()
        val sender = Contact("Bernd", listOf(dropURL), QblECKeyPair().pub)

        val dropMessage = DropMessage(sender, TEST_MESSAGE, TEST_MESSAGE_TYPE)
        assertNull(dropMessage.dropMessageMetadata)

        val gson = DropMessageGson.create()
        val json = gson.toJson(dropMessage)
        assertNotNull(json)

        val receivedDropMessage = gson.fromJson(json, DropMessage::class.java)
        assertNull(receivedDropMessage.dropMessageMetadata)
        assertEquals(sender.keyIdentifier, receivedDropMessage.senderKeyId)
        assertEquals(dropMessage.dropPayload, receivedDropMessage.dropPayload)
        assertEquals(dropMessage.dropPayloadType, receivedDropMessage.dropPayloadType)
    }

    companion object {

        val TEST_MESSAGE = "baz"
        val TEST_MESSAGE_TYPE = "test_message"
    }
}
